src/Database/LSMTree/Internal/CRC32C.hs

```diff
@@ -0,0 +1,220 @@
+    withFile fs file ReadMode (\h -> go h initialCRC32C)
+  where
+    go :: Handle h -> CRC32C -> m CRC32C
+    go !h !crc = do
+      bs <- hGetSome fs h 65504  -- 2^16 - 4 words overhead
+      if BS.null bs
+        then return crc
+        else go h (updateCRC32C bs crc)
+
+
+{- | $checksum-files
+We use @.checksum@ files to help verify the integrity of on disk snapshots.
+Each .checksum file lists the CRC-32C (Castagnoli) of other files. For further
+details see @doc/format-directory.md@.
+
+The file uses the BSD-style checksum format (e.g. as produced by tools like
+@md5sum --tag@), with the algorithm name \"CRC32C\". This format is text,
+one line per file, using hexedecimal for the 32bit output.
+
+Checksum files are used for each LSM run, and for the snapshot metadata.
+
+Typical examples are:
+
+> CRC32C (keyops) = fd040004
+> CRC32C (blobs) = 5a3b820c
+> CRC32C (filter) = 6653e178
+> CRC32C (index) = f4ec6724
+
+Or
+
+> CRC32C (snapshot) = 87972d7f
+-}
+
+type ChecksumsFile = Map ChecksumsFileName CRC32C
+
+-- | File names must not include characters @'('@, @')'@ or @'\n'@.
+--
+newtype ChecksumsFileName = ChecksumsFileName BSC.ByteString
+  deriving (Eq, Ord, Show)
+
+data ChecksumsFileFormatError = ChecksumsFileFormatError FsPath BSC.ByteString
+  deriving Show
+
+instance Exception ChecksumsFileFormatError
+
+readChecksumsFile :: MonadThrow m
+                  => HasFS m h -> FsPath -> m ChecksumsFile
+readChecksumsFile fs path = do
+    str <- withFile fs path ReadMode (\h -> hGetAll fs h)
+    case parseChecksumsFile (BSL.toStrict str) of
+      Left  badline   -> throwIO (ChecksumsFileFormatError path badline)
+      Right checksums -> return checksums
+
+writeChecksumsFile :: MonadThrow m
+                   => HasFS m h -> FsPath -> ChecksumsFile -> m ()
+writeChecksumsFile fs path checksums =
+    withFile fs path (WriteMode MustBeNew) $ \h -> do
+      _ <- hPutAll fs h (formatChecksumsFile checksums)
+      return ()
+
+parseChecksumsFile :: BSC.ByteString -> Either BSC.ByteString ChecksumsFile
+parseChecksumsFile content =
+    case partitionEithers (parseLines content) of
+      ([], entries)    -> Right $! Map.fromList entries
+      ((badline:_), _) -> Left badline
+  where
+    parseLines = map (\l -> maybe (Left l) Right (parseChecksumFileLine l))
+               . filter (not . BSC.null)
+               . BSC.lines
+
+parseChecksumFileLine :: BSC.ByteString -> Maybe (ChecksumsFileName, CRC32C)
+parseChecksumFileLine str0 = do
+    guard (BSC.take 8 str0 == "CRC32C (")
+    let str1 = BSC.drop 8 str0
+    let (name, str2) = BSC.break (==')') str1
+    guard (BSC.take 4 str2 == ") = ")
+    let str3 = BSC.drop 4 str2
+    guard (BSC.length str3 == 8 && BSC.all isHexDigit str3)
+    let !crc = fromIntegral (hexdigitsToInt str3)
+    return (ChecksumsFileName name, CRC32C crc)
+
+isHexDigit :: Char -> Bool
+isHexDigit c = (c >= '0' && c <= '9')
+            || (c >= 'a' && c <= 'f') --lower case only
```

Comment:
There are existing functions for dealing with hex digits in `Data.Char`, although they also accept upper case `A-F` and work on `Int`, so I see how there are advantages to having our own versions here. Or is there even another reason?

Reply:
Prejudice that all Data.Char property functions are slow table-driven implementations that work for all of Unicode.

However you prompted me to look it up and actually `Data.Char.isHexDigit` is ASCII only and uses a clever implementation. So I change my mind :-)

Reply:
On second thoughts, I might not bother since `hexdigitToWord` would also need to be extended to support upper case hex digits.

Reply:
It's fine either way I think. `digitToInt` is pretty similar to `hexDigitToWord`, we'll just need to do one extra `Int -> Word` conversion instead of reusing the one from the guard (which doesn't look like it will be optimised away?). Probably not an issue for what we're doing with it.